sdk/java/pulumi/src/main/java/com/pulumi/test/internal/PulumiTestInternal.java

```diff
@@ -42,4 +147,274 @@ public CompletableFuture<TestResult> runTestAsync(
+         * @see PulumiTest.API#withOptions(TestOptions)
+         */
+        public Builder options(TestOptions options) {
+            this.options = requireNonNull(options);
+            return this;
+        }
+
+        /**
+         * @param mocks the {@link com.pulumi.deployment.internal.Monitor} mocks to use
+         * @return a Pulumi test program entrypoint with given {@link TestOptions}
+         * @see PulumiTest.API#withMocks(Mocks)
+         */
+        public Builder mocks(Mocks mocks) {
+            this.mocks = requireNonNull(mocks);
+            return this;
+        }
+
+        public PulumiTestInternal.Builder standardLogger(Logger standardLogger) {
+            this.standardLogger = requireNonNull(standardLogger);
+            return this;
+        }
+
+        /**
+         * Set configuration to use for this test
+         *
+         * @param allConfig        the configuration key-value map
+         * @param configSecretKeys the secret key names
+         * @return this {@link PulumiTestInternal.Builder}
+         */
+        public PulumiTestInternal.Builder config(Map<String, String> allConfig, Set<String> configSecretKeys) {
+            return internalConfig(new DeploymentImpl.Config(
+                    ImmutableMap.copyOf(allConfig), ImmutableSet.copyOf(configSecretKeys)
+            ));
+        }
+
+        /**
+         * Set configuration to use for this test
+         *
+         * @param allConfig the configuration key-value map
+         * @return this {@link PulumiTestInternal.Builder}
+         */
+        public PulumiTestInternal.Builder config(Map<String, String> allConfig) {
+            return internalConfig(new DeploymentImpl.Config(
+                    ImmutableMap.copyOf(allConfig), ImmutableSet.of()
+            ));
+        }
+
+        private PulumiTestInternal.Builder internalConfig(DeploymentImpl.Config config) {
+            this.config = requireNonNull(config);
+            return this;
+        }
+
+        /**
+         * Modify the deployment factory used in the test
+         *
+         * @param deploymentFactory the factory for
+         * @return this {@link Builder}
+         */
+        public Builder deploymentFactory(
+                Function<DeploymentImpl.DeploymentState, DeploymentInternal> deploymentFactory
+        ) {
+            this.deploymentFactory = requireNonNull(deploymentFactory);
+            return this;
+        }
+
+        /**
+         * @return a {@link PulumiTestInternal} instance created from this {@link Builder}
+         */
+        public PulumiTestInternal build() {
+            if (this.options == null) {
+                this.options = TestOptions.Empty;
+            }
+            if (this.standardLogger == null) {
+                this.standardLogger = defaultLogger();
+            }
+            // FIXME: this instance is different then the one in DeploymentState
```

Comment:
some unfortunate consequences of the current way of initializing the Deployment, we'll address those separately